app/workers/user_cleaner.rb

```diff
@@ -0,0 +1,121 @@
+  #   we have more users with a deletion date lying in the past than
+  #   MAX_DELETIONS_PER_RUN. However, we don't send an additional mail once
+  #   the user is deleted, so this shouldn't be a problem.
+  # - Despite that there cannot be more than MAX_DELETIONS_PER_RUN users with the
+  #   same deletion date, warning mails might be sent on the same date to users
+  #   with varying deletion dates, since the 40-, 14-, 7- and 2-day warning mails
+  #   can overlap temporally.
+  MAX_DELETIONS_PER_RUN = 50
+
+  # Returns all users who have been inactive for 6 months,
+  # i.e. their last sign-in date is more than 6 months ago.
+  def inactive_users
+    User.where("last_sign_in_at < ?", 6.months.ago)
+  end
+
+  # Sets the deletion date for inactive users and sends an initial warning mail.
+  #
+  # This method finds all inactive users whose deletion date is nil (not set yet)
+  # and updates their deletion date to be 40 days from the current date.
+  #
+  # The maximum number of deletion dates set in one run is limited by
+  # MAX_DELETIONS_PER_RUN.
+  def set_deletion_date_for_inactive_users
+    inactive_users.where(deletion_date: nil)
+                  .limit(MAX_DELETIONS_PER_RUN)
+                  .find_each do |user|
+      user.update(deletion_date: Date.current + 40.days)
+      UserCleanerMailer.with(user: user).pending_deletion_email(40).deliver_later
+    end
+  end
+
+  # Unsets the deletion date for users who have been active recently.
+  #
+  # This method finds all users whose deletion date is set and unsets it if the
+  # user has been active in the last 6 months.
+  #
+  # Note that this method just serves as a safety measure. The deletion date
+  # should be unset after every successful user sign-in, see the Warden callback
+  # in `config/initializers/after_sign_in.rb`. If for some reason, the callback
+  # does not work, this method will prevent active users from being deleted
+  # as a last resort.
+  def unset_deletion_date_for_recently_active_users
+    inactive_users_cached = inactive_users
+
+    User.where.not(deletion_date: nil).find_each do |user|
+      user.update(deletion_date: nil) unless inactive_users_cached.include?(user)
+    end
+  end
+
+  # Deletes all users whose deletion date is in the past or present.
+  #
+  # Technically, there should never be users with a deletion date in the past
+  # since the cron job is run daily and should delete users on the day of their
+  # deletion date. Should the cron job not run for some reason, we also delete
+  # users with a deletion date in the past via this method.
+  #
+  # The deletion date for the users must have been set beforehand by calling
+  # `set_deletion_date_for_inactive_users`.
+  def delete_users_according_to_deletion_date!
+    num_deleted_users = 0
+
+    User.where("deletion_date <= ?", Date.current).find_each do |user|
+      next unless user.generic?
+
+      user.destroy
+      num_deleted_users += 1
+    end
```

Comment:
Also, shouldn't we possibly send a mail that confirms the deletion of the account if the user was destroyed?

Reply:
Added in d970c4fbcf